### PR TITLE
[CUDA] Add histogram and histogramdd kernels

### PR DESCRIPTION
**Pull request by zupengwang**
Adds float32 and float64 CUDA implementations of `torch.histogram` and `torch.histogramdd`, covering equal-width and explicit bin edges, multidimensional inputs, weights, density normalization, and noncontiguous inputs.

Histograms accumulate on the GPU using shared memory when they fit and global atomics otherwise. Automatic range inference transfers only per-dimension extrema to the CPU. Existing frontend validation and output handling are reused.

Validation:

- Source build and precommit `spin lint`: passed.
- CPU histogram tests: 53 passed, 1 skipped.
- RTX 4090: 51 CUDA regression tests passed; OpInfo checks: 9 passed, 8 skipped.
- Compute Sanitizer: 30 targeted tests passed, zero memory errors.
- All 20 benchmark cases passed numerical comparisons.

Example timings for one million float32 3D samples with 32 bins per dimension, measured on RTX 4090. The fallback copies inputs to CPU and outputs back to GPU, using two CPU threads. Values are medians of seven measurements after three warmups.

| Input distribution | CUDA | CPU round trip |
| --- | ---: | ---: |
| Random samples, equal-width bins | 0.186 ms | 16.320 ms |
| All samples in one bin | 1.537 ms | 16.198 ms |

Floating-point accumulation is nondeterministic. FP16/BF16 are unsupported; ROCm, CUDA Graphs, and compiled execution were not tested.

AI-assisted.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -20548,7 +20548,7 @@ def sample_inputs_abs(op_info, device, dtype, requires_grad, op_kwargs=None, **k
                     supports_rhs_python_scalar=False),
     OpInfo('histogram',
            dtypes=floating_types(),
-           dtypesIfCUDA=_dispatch_dtypes(),  # histogram is only implemented on CPU
+           dtypesIfCUDA=floating_types(),
```
Comment (Skylion007) [inline torch/testing/_internal/common_methods_invocations.py]: You can drop the whole dtypeIfCUDA since the default sentinel will just feed from dtypes

**Comment by Skylion007** [comment]
Doesn't NVidia cub which we already link have these kernels? Can we just add dispatch to these optimized kernels?

**Comment by Skylion007** [comment]
@claude should we use Nvidia cub/thrust kernels for these histogram kernels? 

How would you do so?

**Comment by Skylion007** [comment]
CUB DeviceHistogram is an excellent fit for ordinary 1-D torch.histogram, especially the equal-width case.

**Comment by Skylion007** [comment]
Okay actually I don't see an easy way to avoid a copy when writing out the CUB result, hmm...

**Comment by Skylion007** [comment]
Nice improvement. I looked into whether CUB could replace the custom histogram kernel, but I do not think it is a particularly clean fit here: `DeviceHistogram` is primarily a counter histogram, while this implementation needs float/double output and arbitrary weighted accumulation, so using CUB would likely require a temporary integer histogram plus a conversion (and still would not cover the general `histogramdd` case).

One optimization that seems more worthwhile is a dedicated 1-D path. Currently ordinary `torch.histogram` appears to go through the generic `histogramdd` machinery, including the runtime dimension loop and construction/copy of `edge_offsets`. For `dimensions == 1`, a specialized kernel could take the edge tensor directly and avoid the offsets metadata entirely. It might also be worth compile-time specializing weighted vs. unweighted; the latter could potentially use warp-aggregated atomics for high-contention distributions (the all-one-bin benchmark seems to be where the current kernel loses the most performance).

For the linear/equal-width case, could we also avoid consulting the materialized edge tensor during bin lookup and instead pass `(left, right, num_bins)` to the kernel? The edges can still be materialized for the returned API value, but computing the bin analytically would avoid edge loads in the hot path. Obviously the exact right-edge/rounding semantics would need to be preserved.

I would be interested in seeing a 1-D benchmark as well, since that is probably the common case and is where the generic `histogramdd` setup overhead is least necessary.

**Comment by zupengwang** [comment]
@Skylion007 Updated in ca51776. Thanks for the suggestions. This update adds a dedicated 1-D specialization that uses the edge tensor directly, eliminating the offsets allocation and transfer. It also adds compile-time weighted/unweighted dispatch, shared-memory replicas, warp aggregation for shared and global accumulation, and 32-bit indexing where safe with a 64-bit fallback. Both redundant `dtypesIfCUDA` entries are removed.

I retained the checks against the returned edges for equal-width binning: narrow float32 ranges can produce repeated rounded edges, where formula-only bin assignment disagrees with the returned boundaries. Regression tests cover this case and the rightmost edge.

For one million float32 samples on RTX 4090, compared with the original PR implementation:

| Workload, unweighted | Before | After | Speedup |
|---|---:|---:|---:|
| 1-D random, 100 bins | 0.0611 ms | 0.0388 ms | 1.58x |
| 1-D all in one bin, 100 bins | 0.1058 ms | 0.0398 ms | 2.66x |
| 3-D all in one bin, 32 bins/dimension | 1.5358 ms | 0.1743 ms | 8.81x |
| 3-D random, 32 bins/dimension | 0.1718 ms | 0.1722 ms | ~1.00x |

These are synchronized wall-clock measurements including host setup, using one CPU thread and A-B-B-A runs. Both versions use CUDA 13.0/SM86 builds. All 96 benchmark configurations passed numerical checks. Some random 3-D float64 cases still regress: up to 6.1% on 4090 and 11.9% on 3090, so the improvement is workload-dependent.

AI-assisted.

**Comment by Skylion007** [comment]
@claude investigate why there is regression for the 3D histogram cases and see if they can be fixed easily

File: aten/src/ATen/native/cuda/HistogramKernel.cu [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        return pos + 1;
+      }
+    }
+    // Fall back for overflow or repeated edges caused by rounding linspace.
+  }
+
+  index_t first = 0;
+  index_t last = num_edges;
+  while (first < last) {
+    const index_t mid = first + (last - first) / 2;
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/HistogramKernel.cu]:
```suggestion
    const index_t mid = cuda::std::midpoint(first, last);

**Comment by Skylion007** [comment]
Also merging this midpoint utility you should probably use: https://github.com/pytorch/pytorch/pull/196356

**Comment by zupengwang** [comment]
@Skylion007 Updated in d78f46cb. Thanks for the follow-up and the midpoint pointer.

The replica budget now accounts for `sharedMemPerMultiprocessor / 4`, while preserving the existing shared/global dispatch threshold. This update also uses `C10_WARP_SIZE` and the midpoint helper from #196356. Its unchanged header hunk is included because this branch's base predates it.

Profiling confirmed that limiting replicas helps small shared-memory histograms: for the measured 8³-bin float64 weighted workload on RTX 3090, achieved occupancy increased from 41.3% to 63.0%. However, the 32³-bin cases use global atomics for both dtypes, so replica sizing cannot explain their earlier regression.

For equal-width binning, the position estimate now uses float arithmetic, with subtraction performed in the input dtype. Candidate bins are still checked against the original edges, with the existing binary-search fallback. New tests cover extreme float64 ranges, repeated edges, and more than 2²⁴ bins. Weighted accumulation uses a shuffle and vote to identify uniform warps; unweighted accumulation retains full peer-group aggregation.

Compared with ca51776, for one million random float64 samples:

| Workload | RTX 3090, before → after | Speedup | RTX 4090, before → after | Speedup |
| --- | --- | --- | --- | --- |
| 1-D, 100 equal-width bins, unweighted | 0.1213 → 0.0864 ms | 1.40× | 0.0734 → 0.0625 ms | 1.18× |
| 3-D, 32³ equal-width bins, weighted | 0.3762 → 0.2517 ms | 1.49× | 0.2329 → 0.1776 ms | 1.31× |
| 3-D, 8³ explicit bins, weighted | 0.2544 → 0.2269 ms | 1.12× | 0.1875 → 0.1696 ms | 1.11× |

These are synchronized wall-clock measurements including host setup, using one CPU thread and A-B-B-A runs. Both versions use CUDA 13.0/SM86 builds. Longer warmups reduced frequency-related variation observed in earlier measurements; clocks were recorded but not fixed. All 192 standard and 28 mixed-hotspot configurations passed numerical checks on each GPU.

Performance remains workload-dependent. Across the full matrix, the largest observed regressions were 5.9%/5.2% in call latency and 6.9%/4.0% in Graph replay on 3090/4090, respectively.

Validation: 257 CUDA tests passed on two RTX 3090s; 229 passed on RTX 4090, with 28 multi-GPU cases skipped. Memcheck and racecheck each passed 253 cases with zero errors. The suite includes CUDA Graphs, streams, and multi-GPU coverage. Precommit lint also passed.

AI-assisted.

**Comment by Skylion007** [comment]
@claude review this PR

**Comment by Skylion007** [comment]
Can we also use the shape / error checking from a unified header utility?

**Comment by zupengwang** [comment]
@Skylion007 Updated in 532c274. Thanks for the follow-up and the detailed review.

The common frontend now uses `checkScalarType`, `checkDim`, `checkSize`, and `check_and_update_common_device`. Weight device validation is shared across backends, and the redundant CUDA checks are removed.

I also addressed the CPU/CUDA inconsistency for repeated rounded edges: CPU lookup now falls back to a full search when the local window misses. The CPU skip is removed, with NumPy comparisons covering both dtypes, weights, and 1-D/3-D inputs. Per-dimension scales are cached to avoid repeated division.

The docstrings now explain the synchronization and CUDA Graph restriction when a nonempty CUDA input needs range inference. Capture-error tests run in-process and verify that a subsequent explicit-range capture succeeds. New tests straddle the device's actual shared-memory limit. This update also uses portable shuffle wrappers and fixed-block launch bounds, and documents the lookup and backend fallbacks.

For inferred ranges, `aminmax_out` writes directly into one contiguous buffer before copying to the CPU. Nsight confirmed that D2H copies and stream synchronizations were halved, without additional kernel launches. Per call, this is 2 → 1 for `histogram` and 4 → 2 for `histogramdd`, whose frontend currently infers the range twice.

On RTX 3090, compared with d78f46c:

| Inferred-range workload | Before | After | Speedup |
| --- | ---: | ---: | ---: |
| float32, 1-D, 256 samples, unweighted | 0.09163 ms | 0.07731 ms | 1.185x |
| float32, 1-D, 1M samples, unweighted | 0.09926 ms | 0.08667 ms | 1.145x |
| float64, 3-D, 1M samples, weighted | 0.48341 ms | 0.45980 ms | 1.051x |

These are synchronized wall-clock measurements including host setup, using one CPU thread and A-B-B-A runs with CUDA 13.0/SM86 builds. Clocks were recorded but not locked. All 16 inferred-range and 80 fixed-range/explicit-edge configurations passed numerical checks. Fixed-range call latency was approximately unchanged at the median; the largest observed call/Graph latency increases were 3.37%/3.96%. The largest CPU small-input increase was 1.89 us (8.24%). Concurrent tasks made the 4090 timings unreliable, so I am excluding them from performance claims.

Validation:
- CUDA: 327 passed on two RTX 3090s; 299 passed on RTX 4090, with 28 multi-GPU cases skipped.
- CPU histogram/histc: 112 passed, 230 skipped, plus three reference/error-handling tests.
- Memcheck and racecheck each passed the 325-case suite using the final CUDA artifact, followed by a 14-case extreme-range rerun covering two newly added instances. Zero errors or race warnings.
- OpInfo and precommit lint passed.

AI-assisted.

**Comment by zupengwang** [comment]
@Skylion007 Could you take another look at 532c274 when you have a chance? This update uses the shared shape and error-checking utilities, fixes the CPU/CUDA discrepancy for repeated bin edges, and adds the inferred-range synchronization documentation and CUDA Graph regression coverage. The changes and validation results are summarized in [the update above](https://github.com/pytorch/pytorch/pull/196280#issuecomment-5606649158). Thanks!
